### PR TITLE
Fix evaluation of constant expression containing mapped properties

### LINKED ISSUE #639: Filter not working when passing expand entity property
When running a query that contains filter with a link entity property it returns an error message:

Microsoft.OData.ODataException: 'The type 'WebAppOdata.Account' is not supported when converting to a URI literal.'

The code looks like this:

            // get service
            IEnumerable<Billing_Service> billingserviceForSubscription = await odataClient.For<Billing_Service>()
                .Select(x => new { x.Id, x.Name, x.Account }).Expand(x => new { x.Account }).Select(x => new { x.Account.Id, x.Account.Name }).FindEntriesAsync();
            // get current account
            Account current_account = billingserviceForSubscription.First().Account;
            // get current service
            Billing_Service current_billing_serice = billingserviceForSubscription.First();

            Guid accID = current_account.Id;
            Guid billServiceID = current_billing_serice.Id;

            // get subscriptions
            var subscriptions = await odataClient.For<Suscription>()
                .Expand(x => x.Account).Expand(x => x.Billing_Service)
                .Filter(x => x.Account.Id == ((Guid)current_account.Id) && x.Billing_Service.Id == billServiceID).GetCommandTextAsync();

The Billing Service, Subscription and Account are CRM entities.

The definition of the Entities look like this:

    public class Account
    {
        [Extensions.Column(Name = "xxxxx")]
        public Guid Id { get; set; }
        [Extensions.Column(Name = "name")]
        public string Name { get; set; }
    }

    [Table("xxxxxxx")]
    public class Billing_Service
    {
        [Extensions.Column(Name= "xxxxx")]
        public Account Account { get; set; }

        [Extensions.Column(Name = "xxxxx")]
        public Billing_Period_Type Billing_Period_Type { get; set; }

        [Extensions.Column(Name = "xxxxx")]
        public Guid Id { get; set; }

        [Extensions.Column(Name= "statuscode")]
        public string StatusCode { get; set; }
        [Extensions.Column(Name= "xxxxx")]
        public string Name { get; set; }

        [Extensions.NotMapped]
        public BillingEnabled BillingEnabled { get; set; }

        [Extensions.Column(Name = "xxxxx")]
        public long BillingEnabledOS
        {
            get
            {
                return (long)BillingEnabled;
            }
            set
            {
                BillingEnabled = (BillingEnabled)value;
            }
        }
    }
    [Table("xxxxx")]
    public class Suscription
    {
        [Extensions.Column(Name= "xxxxx")]
        public Guid Id { get; set; }
        [Extensions.Column(Name= "xxxxx")]
        public string Name { get; set; }
        [Extensions.Column(Name= "xxxxx")]
        public Account Account { get; set; }
        [Extensions.Column(Name= "xxxxx")]
        public Billing_Service Billing_Service { get; set; }
    }

    [Serializable]
    public enum BillingEnabled
    {
        Yes = 400_000_001,
        No = 400_000_000
    }

When you pass a hard coded variable in the query filter it works.

            // get subscriptions
            var subscriptions = await odataClient.For<Suscription>()
                .Expand(x => x.Account).Expand(x => x.Billing_Service)
                .Filter(x => x.Account.Id == accID && x.Billing_Service.Id == billServiceID).GetCommandTextAsync();

I also tried to run the query with another class that has same properties but does NOT have map attributes and that also works with filter.

            Account2 acc2 = current_account;

            var subscriptions = await odataClient.For<Suscription>()
                .Expand(x => x.Account).Expand(x => x.Billing_Service)
                .Filter(x => x.Account.Id == acc2.Id && x.Billing_Service.Id == billServiceID).GetCommandTextAsync();

The class looks like this:

    public class Account2
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
    }

### PR DESCRIPTION
**Pull request by maxbruecken**
This pull request fix evaluation of constant expressions containing mapped properties. This change should fix #639

### CODE REVIEW

File: src/Simple.OData.Client.Core/Expressions/ODataExpression.Linq.cs [hunk 1]
```diff
@@ -194,9 +197,8 @@ private static ODataExpression ParseConstantExpression(Expression expression, st
                 }
                 else
                 {
-                    return new ODataExpression(EvaluateConstValue(
-                        constExpression.Type, constExpression.Value,
-                        memberNames == null ? new List<string>() : memberNames.Split('.').ToList()));
+                    return new ODataExpression(EvaluateConstValue(constExpression.Value,
+                        members == null ? new Stack<MemberInfo>() : members));
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Expressions/ODataExpression.Linq.cs:201]:
```suggestion
                    return new ODataExpression(EvaluateConstValue(constExpression.Value,
                        members ?? new Stack<MemberInfo>()));
```

File: src/Simple.OData.Client.Core/Expressions/ODataExpression.Linq.cs [hunk 2]
```diff
@@ -73,24 +73,27 @@ private static ODataExpression ParseMemberExpression(Expression expression, stri
             }
             else
             {
-                // NOTE: Can't support ITypeCache here as we might be dealing with dynamic types/expressions
-                var memberName = memberExpression.Member.GetMappedName();
-
-                memberNames = memberNames == null ? memberName : string.Join(".", memberName, memberNames);
+                if (memberChain == null)
+                    memberChain = new Stack<MemberInfo>();
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Expressions/ODataExpression.Linq.cs:77]:
```suggestion
                memberChain ??= new Stack<MemberInfo>();
```

**Review by watfordgnf — COMMENTED** [review]
If we're updating the language used we may as well simplify some of the other usages.

File: src/Simple.OData.Client.Core/Simple.OData.Client.Core.csproj [hunk 3]
```diff
@@ -6,6 +6,7 @@
     <OutputPath>$(SolutionDir)/build/$(Configuration)/$(AssemblyName)/$(TargetFramework)/</OutputPath>
     <OutDir>$(OutputPath)</OutDir>
     <Decription></Decription>
+    <LangVersion>default</LangVersion>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Simple.OData.Client.Core.csproj:9]: I'm actually surprised you could use pattern matching switches without this before!

Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Simple.OData.Client.Core.csproj:9]: I'm not sure but is pattern matching introduced in C# 7.0?

Comment (watfordgnf) [reply src/Simple.OData.Client.Core/Simple.OData.Client.Core.csproj:9]: Looks like you're correct for the switch statement:
https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.0/pattern-matching.md

Looks like it was 8.0 that added the expression switches.

**Pull request merged**